CAIPs/caip-stream-sync.md

```diff
@@ -0,0 +1,193 @@
+  separatorValue nullable Bytes
+  streamFilter nullable Bytes
+  ctrlFilter nullable Bytes
+} representation tuple
+
+type Tips struct {
+  streamid Bytes
+  tips [CID]
+} representation tuple
+```
+
+*Properties:*
+
+* `tips` - all the tips that the node has for this set
+  * `streamid` - the id of the corresponding tips
+  * `tips` - a list of latest tips
+
+* `separatorKey`, `separatorValue`, `streamFilter`, and `controllerFilter` - same as in *SubscribeSet*
+
+#### Synchronizing Events
+
+The following two protocol messages are used to synchronize event data between peers.
+
+**WantStream**
+
+Request to stream events from another node given a streamId.
+
+```ipldsch
+type WantStream struct {
+  streamid Bytes
+  knownTips [CID] // the CID(s) of the event to start the streaming from
+  offset nullable Integer // number of bytes already received
+} representation tuple
+```
+
+*Properties:*
+
+* `streamid` - the streamId being requested
+* `knownTips` - the CIDs of the tips that the requester already has locally
+* `offset` - number of bytes already recieved on top of the known tip
+
+**StreamChunk**
+
+When a node receives a *WantStream* it sends an event chunk as a response. The event bytes are sent in verification order (starting with the oldest event, ending with the most recent event).
+
+```ipldsch
+type StreamChunk struct {
+  eventData Bytes
+} representation tuple
+```
+
+*Properties:*
+
+* `eventData` - event data bytes
+
+### DHT usage
+
+In order to find other peers which are likely to already be subscribed to the stream sets your node wants to synchronize the libp2p DHT is utilized. The DHT key used is generated from properties in the *StreamSync* message as follows:
+
+* If the `separatorKey/Value` properties are specified, use `hash(separatorKey | separatorValue)` 
+* If `streamFilter` is a full streamid, use `hash(streamFilter)`
+
+This gives nodes an easy way to find other nodes of either a specific stream or a set of streams as defined by the separator.
+
+If the `streamFilter` only contains a smaller number of bytes it may be desirable to construct the key as `hash(separatorKey | separatorValue | streamFilter)`. However, since nodes are free to set the streamFilter however they want it might be hard for a large group of nodes to converge unless they coordinate out of band.
+
+### Lifecycle
+
+Four aspects of the protocol lifecycle are described below.
+
+#### Tip Subscription
+
+1. Peer `A` wants to subscribe to stream set `s`
+2. `A` looks up other peers that subscribe using the libp2p DHT, it finds peer `B`
+3. `A` also announces to the DHT that it is subscribed to `s`
+4. `A` sends a `SubscribeSet` message for `s` to `B`
+5. `B` registers that `A` is subscribed to `s` with a ttl of 12h
+6. `B` sends a `HaveTips` message with the tips it has validated as a response to `A`
+7. `A` registers that `B` is subscribed to `s'` (as defined in the `HaveTips` message) with a ttl of 12h
+
+#### Publishing updates
```

Comment:
so it seems like with this strategy, if a new event in a set is published, let's say to node A, then A first tells the subscribed peers that it has a new event, then the subscribes have to come back and request the actual event data, and then A can send the new events.  Should there be a way to skip the first two steps and just have A immediately begin publishing new events to known subscribers as soon as new events come in?